### ISSUE TITLE
Changing the 'Request Instructor Account' form from Google Form to a request processing system (and a GSOC Introduction)

### ISSUE THREAD

**Issue by HirdayGupta** [TEAMMATES/teammates#6976]
## **Introduction:**
Hello! My name is Hirday Gupta and I am High School Senior (Year 12) from Mumbai, India. I plan to apply to TEAMMATES for GSOC'17 and I am extremely excited to contribute to this wonderful community and remain a part of it even after GSOC is over.

I am proficient in JAVA, Python, HTML, CSS(and bootstrap), PHP, Jquery/JavaScript, MySQL, and have experience with iOS and Android App Development. You can find the source code for all of my projects on [my Github page](https://github.com/HirdayGupta). I have also developed an[ iOS App](https://appsto.re/in/-qKbfb.i) and a [JAVA Library](https://sourceforge.net/projects/javaarduinolibrary/). I also like to believe that I am a fast learner and have begun reading and learning about the Google App Engine environment.

## **The Form:**
I have set up a copy of TEAMMATES on my computer and am currently working on _changing the Instructor Account request form from a Google Form to a native form - a feature listed on the Spec as pending._ 

I have finished making the front end of the form (seen in the attached image) and implemented basic functionality such as required fields, and browser's inbuilt email verification. 
<img width="1280" alt="form screenshot" src="https://cloud.githubusercontent.com/assets/18088721/24465654/69d027bc-14cb-11e7-872b-a83cbf7fc503.png">

## **Questions**
While developing the backend, I had some questions about features required in the form and also about the platform in general.

 1. Using javascript/jquery:
    - I ran into some problems while trying to link a .js file to my the 'request.jsp' file to implement additional form functionality.
    - I created a file called 'request.js' and stored it in the _main/webapp/js_ directory. Then, in the 'request.jsp' file, I included `<script type="text/javascript" src="/js/request.js"></script>` before the `<t:staticPage>` declaration. 
    - However, this did not seem to work as even the following basic code did not run. (No alert box showed on loading the page).
        `$( document ).ready(function() {`
            `alert( "ready!" );`
        `});`
    - I also tried to mimic the way the remaining JSP files included js files, by adding `<%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>` and enclosing the <script> tags in `<c:set var="jsIncludes"></c:set>` but that didn't work either. However, I am also looking to understand this method of inclusion and would like to request being pointed to any documentation or an explanation.
    - I would like to request for guidance on the right way to link JSP files with js script files (or to point out if I'm just doing something stupid :p)

2. Regarding the storage of the data entered:
    - I was wondering if you would like the data entered in this form to be stored in the GAE Datastore or simply emailed to you with a summary of the form fields.
    - If you would like this data to be stored in the GAE Datastore, I would require some guidance regarding the processes involved as I was unable to find an example in the downloaded code.

3. I was a little confused about the email sending infrastructure that is currently set up in TEAMMATES and was wondering:
    - Where can I get a technical description of the email infrastructure currently in use
    - What are your preferences regarding email notifications after the submission of this form (for instance, notification to admin with form field details, confirmation to user, etc)

## **Features to be implemented:**
1. After sorting out the above issues, I will use javascript to implement functionality as in the Google Form which will inform the user if he has left any required form field complete (before and after submission) as well as validate the fields.

2. If the datastore is to be used, I will use it to check if an instructor has already requested a form, and depending on its status (an additional field I will include in the datastore), if he has been granted an account, the website will redirect to the login page, or if his request is pending a reminder will be sent to the admin.

3. Lastly, the admin will also be notified by email every time the form is submitted.

I would like to add that one of my main reasons for participating in GSOC is to learn about the industry standards regarding code quality and structure. Simply by observing the structure of this project, and reading through the code - including the succinct documentation - I am already learning so much. I am extremely excited to contribute to this project and continue the learning process.

## **TL;DR**
1. View the screenshot to see the native form replacing the Google Form for Requesting an Instructor Account. This was listed as pending on the Spec.
2. Read the "Questions" section for the issues.

**Comment by HirdayGupta** [comment]
@damithc I have posted my queries here as per your instructions on email. I would also like to ask if the change I am proposing here is advisable to complete before GSOC or as a part of my proposal?

**Comment by damithc** [comment]
@HirdayGupta is this your proposal for GSoC? In that case I recommend you choose a project idea listed in our own list. Also, those with `high priority` ratings are more likely to be selected.

**Comment by HirdayGupta** [comment]
@damithc This is not my proposal for GSOC but merely a first contribution I would like to make as an introduction to myself before my proposal. I am currently drafting a proposal for the 'Usability Hero' project idea. I was asking for your opinion on whether this is a viable first contribution or if I should just include this as part of my proposal.

Either way, I would be extremely grateful for answers to the queries in the *Questions* section of my post.

**Comment by damithc** [comment]
I see @HirdayGupta I'm impressed that you attempted to add this whole new feature as your first contribution. however, this is a bit too big for a new contributor, and also not a feature we are looking to add at the moment (and therefore we are not willing to spend mentor resources on).
Perhaps you can tackle a `d.FirstTimers` or a `d.Contributors` issue instead?

**Comment by HirdayGupta** [comment]
Thank you for the response @damithc 
I have already solved a `d.FirstTimers` issue [here](https://github.com/TEAMMATES/teammates/pull/6992) and the edit has been merged into the master.  I was suggesting this as I felt that fixing a typo was not enough contribution. 

Anyhow, thank you for your feedback. I will now work on my draft and send it in for review as soon as possible. Thanks!

**Comment by damithc** [comment]
> I have already solved a d.FirstTimers issue here and the edit has been merged into the master. I was suggesting this as I felt that fixing a typo was not enough contribution.

That's true. Time to move to a `d.Contributor` issue which are slightly harder. :-)

**Comment by damithc** [comment]
Closing to focus on higher priority issues. Will be reopened later.
